src/dbnode/encoding/proto/encoder.go

```diff
@@ -0,0 +1,820 @@
+				encErrPrefix, customField.fieldNum)
+		}
+
+		customEncoded := true
+		switch {
+		case isCustomFloatEncodedField(customField.fieldType):
+			if err := enc.encodeTSZValue(i, iVal); err != nil {
+				return err
+			}
+		case isCustomIntEncodedField(customField.fieldType):
+			if err := enc.encodeIntValue(i, iVal); err != nil {
+				return err
+			}
+		case customField.fieldType == cBytes:
+			if err := enc.encodeBytesValue(i, iVal); err != nil {
+				return err
+			}
+		default:
+			customEncoded = false
+		}
+
+		if customEncoded {
+			// Remove the field from the message so we don't include it
+			// in the proto marshal.
+			m.ClearFieldByNumber(customField.fieldNum)
+		}
+	}
+	enc.hasEncodedFirstSetOfCustomValues = true
+
+	return nil
+}
+
+func (enc *Encoder) encodeTSZValue(i int, iVal interface{}) error {
+	var (
+		val         float64
+		customField = enc.customFields[i]
+	)
+	switch typedVal := iVal.(type) {
+	case float64:
+		val = typedVal
+	case float32:
+		val = float64(typedVal)
+	default:
+		return fmt.Errorf(
+			"%s found unknown type in fieldNum %d", encErrPrefix, customField.fieldNum)
+	}
+
+	if !enc.hasEncodedFirstSetOfCustomValues {
+		enc.encodeFirstTSZValue(i, val)
+	} else {
+		enc.encodeNextTSZValue(i, val)
+	}
+
+	return nil
+}
+
+func (enc *Encoder) encodeIntValue(i int, iVal interface{}) error {
+	var (
+		signedVal   int64
+		unsignedVal uint64
+		customField = enc.customFields[i]
+	)
+	switch typedVal := iVal.(type) {
+	case uint64:
+		unsignedVal = typedVal
+	case uint32:
+		unsignedVal = uint64(typedVal)
+	case int64:
+		signedVal = typedVal
+	case int32:
+		signedVal = int64(typedVal)
+	default:
+		return fmt.Errorf(
+			"%s found unknown type in fieldNum %d", encErrPrefix, customField.fieldNum)
+	}
+
+	if isUnsignedInt(customField.fieldType) {
+		if !enc.hasEncodedFirstSetOfCustomValues {
+			enc.encodeFirstUnsignedIntValue(i, unsignedVal)
+		} else {
+			enc.encodeNextUnsignedIntValue(i, unsignedVal)
+		}
+	} else {
+		if !enc.hasEncodedFirstSetOfCustomValues {
+			enc.encodeFirstSignedIntValue(i, signedVal)
+		} else {
+			enc.encodeNextSignedIntValue(i, signedVal)
+		}
+	}
+
+	return nil
+}
+
+func (enc *Encoder) encodeBytesValue(i int, iVal interface{}) error {
+	customField := enc.customFields[i]
+	currBytes, ok := iVal.([]byte)
+	if !ok {
+		currString, ok := iVal.(string)
+		if !ok {
+			return fmt.Errorf(
+				"%s found unknown type in fieldNum %d", encErrPrefix, customField.fieldNum)
+		}
+		currBytes = []byte(currString)
+	}
+
+	var (
+		hash             = murmur3.Sum64(currBytes)
+		numPreviousBytes = len(customField.bytesFieldDict)
+		lastHashIdx      = numPreviousBytes - 1
+	)
+	if numPreviousBytes > 0 && hash == customField.bytesFieldDict[lastHashIdx] {
+		// No changes control bit.
+		enc.stream.WriteBit(opCodeNoChange)
+		return nil
+	}
+
+	// Bytes changed control bit.
+	enc.stream.WriteBit(opCodeChange)
+
+	for j, prevHash := range customField.bytesFieldDict {
+		if hash == prevHash {
```

Comment:
Here too, need to actually check values match with bytes.Equal or something (subslice as we talked about back into where it was stored).

Reply:
same reply